### PR TITLE
Emit diagnostic for privately uninhabited uncovered witnesses.

### LINKED ISSUE #104034: `#[feature(exhaustive_patterns)]` thinks irrefutable pattern is refutable when enum contains struct with private fields.
The following code compiles on nightly ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=55ac489a6a9fcd1189f52d4d2c254e94)):

```rust
#![feature(exhaustive_patterns, never_type)]

enum Either<A, B> {
    A(A),
    B(Wrapper<B>),
}

struct Wrapper<T>(T);

fn foo() -> Either<(), !> {
    Either::A(())
}

fn main() {
    let Either::A(()) = foo();
}
```

But if the wrapper type is moved into a module to make the field private, then it does not: ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=81d51fa51530ad29918e0d6e078c3070)):

```rust
#![feature(exhaustive_patterns, never_type)]

mod inner {
    pub struct Wrapper<T>(T);
}

enum Either<A, B> {
    A(A),
    B(inner::Wrapper<B>),
}

fn foo() -> Either<(), !> {
    Either::A(())
}

fn main() {
    let Either::A(()) = foo();
}
```

```
error[E0005]: refutable pattern in local binding: `Either::B(_)` not covered
  --> src/main.rs:17:9
   |
17 |     let Either::A(()) = foo();
   |         ^^^^^^^^^^^^^ pattern `Either::B(_)` not covered
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
note: `Either<(), !>` defined here
  --> src/main.rs:9:5
   |
7  | enum Either<A, B> {
   |      ------
8  |     A(A),
9  |     B(inner::Wrapper<B>),
   |     ^ not covered
   = note: the matched value is of type `Either<(), !>`
help: you might want to use `if let` to ignore the variant that isn't matched
   |
17 |     if let Either::A(()) = foo() { todo!() }
   |     ++                           ~~~~~~~~~~~
```

Making the field `pub` makes it compile again.

Is this expected? I suppose I could understand if it is; the private fields could change to make `Wrapper<!>` actually be constructable and therefore make the pattern refutable. But if that's the case, then I think the compiler should point this out, like `note: the pattern is currently irrefutable, but the type contains private fields which may change in the future to make the pattern refutable`.

If indeed this is expected, is there any way for me as a library author to somehow convince the compiler that I will never change the fields to make the type constructable? I want users of my library to be able to elide match arms when I give them an `Either<A, !>`.

@rustbot label +T-compiler +F-never_type +D-confusing +requires-nightly +S-bug-has-mcve

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/104034

cc @Nadrieril

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 1]
```diff
@@ -479,12 +479,30 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
...
+            if cx.tcx.features().exhaustive_patterns
+                && let Some(witness_1) = witnesses.get(0)
+                && let ty::Adt(adt, substs) = witness_1.ty().kind()
+                && adt.is_enum()
+                && let Constructor::Variant(variant_index) = witness_1.ctor()
+            {
+                let variant = adt.variant(*variant_index);
+                let inhabited = variant.inhabited_predicate(cx.tcx, *adt).subst(cx.tcx, substs);
+                assert!(inhabited.apply(cx.tcx, cx.param_env, cx.module));
+                !inhabited.apply_ignore_module_in_crate(cx.tcx, cx.param_env)
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:494]: I'm slightly confused about the logic here: from reading the doc-comments of these two functions, don't they return the same thing for any module that is not in the current crate?

File: tests/ui/never_type/exhaustive_patterns.stderr [hunk 2]
```diff
@@ -14,6 +14,7 @@ LL | enum Either<A, B> {
 LL |     A(A),
 LL |     B(inner::Wrapper<B>),
    |     - not covered
+   = note: pattern `Either::B(_)` is currently uninhabited, but this variant contains private fields which may change in the future
```
Comment (Nadrieril) [inline tests/ui/never_type/exhaustive_patterns.stderr:17]: Could you add the appropriate magic comments to the corresponding rust file so we don't accidentally regress this?

**Comment by petrochenkov** [comment]
I'm not very familiar with this code, but the change is diagnostic-only and the output looks reasonable.
r=me after addressing https://github.com/rust-lang/rust/pull/111624#discussion_r1195060932 and squashing commits.
@rustbot author

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs (hunk 1 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/check_match.rs:494]: My understanding of what we want the logic to do is: if the first witness is a variant that is uninhabited *in its native crate/module* (or in fact anywhere at all is enough) then we emit a note. It shouldn't be about the current crate

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 3]
```diff
@@ -479,12 +479,30 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
             AdtDefinedHere { adt_def_span, ty, variants }
         };
 
+        // Emit an extra note if the first uncovered witness is
+        // visibly uninhabited anywhere in the current crate.
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:483]: It seems like you changed the logic, could you update this comment to match the new logic?

**Comment by Nadrieril** [comment]
It seems you changed the logic to what I suggested, could you add a test where the secretely uninhabited type is in another crate? `tests/ui/pattern/usefulness/auxiliary/empty.rs` has `SecretlyUninhabitedForeignStruct` you can reuse for tests.

**Pull request merged**